gdnative-derive/src/native_script.rs

```diff
@@ -61,54 +61,112 @@ pub(crate) fn derive_native_class(derive_input: &DeriveInput) -> Result<TokenStr
-
-            let label = config.path.unwrap_or_else(|| format!("{}", ident));
-            quote!({
-                builder.property(#label)
-                    #with_default
-                    #with_hint
-                    #with_usage
-                    .with_ref_getter(|this: &#name, _owner: ::gdnative::object::TRef<Self::Base>| {
-                        #before_get
-                        let res = &this.#ident;
-                        #after_get
-                        res
-                    })
+        let properties = data
+            .properties
+            .into_iter()
+            .map(|(ident, config)| {
+                let with_default = config
+                    .default
+                    .map(|default_value| quote!(.with_default(#default_value)));
+                let with_hint = config.hint.map(|hint_fn| quote!(.with_hint(#hint_fn())));
+                let with_usage = if config.no_editor {
+                    Some(quote!(.with_usage(::gdnative::export::PropertyUsage::NOEDITOR)))
+                } else {
+                    None
+                };
+                // check whether this property type is `Property<T>`. if so, extract T from it.
+                let property_ty = match config.ty {
+                    Type::Path(ref path) => path
+                        .path
+                        .segments
+                        .iter()
+                        .last()
+                        .filter(|seg| seg.ident == "Property")
+                        .and_then(|seg| match seg.arguments {
+                            syn::PathArguments::AngleBracketed(ref params) => params.args.first(),
+                            _ => None,
+                        })
+                        .and_then(|arg| match arg {
+                            syn::GenericArgument::Type(ref ty) => Some(ty),
+                            _ => None,
+                        })
+                        .map(|ty| quote!(::<#ty>)),
+                    _ => None,
+                };
+                // #[property] is not attached on `Property<T>`
+                if property_ty.is_none()
+                // custom getter used
+                && config.get.as_ref().map(|get| !matches!(get, PropertyGet::Default)).unwrap_or(false)
+                // custom setter used
+                && config.set.as_ref().map(|set| !matches!(set, PropertySet::Default)).unwrap_or(false)
+                {
+                    return Err(syn::Error::new(
+                        ident.span(),
+                        "The `#[property]` attribute can only be used on a field of type `Property`, \
+                        if a path is provided for both get/set method(s)."
+                    ));
+                }
+                // if both of them are not set, i.e. `#[property]`. implicitly use both getter/setter
+                let (get, set) = if config.get.is_none() && config.set.is_none() {
```

Comment:
This is what I tried. Test string needs to be adapted too.
I got some error `procedural macro API is used outside of a procedural macro` however; not sure why exactly.

```suggestion
// Attribute is #[property] (or has other arguments which are not relevant here)
                let is_standalone_attribute = config.get.is_none() && config.set.is_none();

                // Attribute is #[property(get)] or #[property(get, set="path")]
                let has_default_getter = matches!(config.get, Some(PropertyGet::Default));

                // Attribute is #[property(set)] or #[property(get="path", set)]
                let has_default_setter = matches!(config.set, Some(PropertySet::Default));

                // Field type is not `Property<T>`
                if property_ty.is_none()
                    && !is_standalone_attribute
                    && (has_default_getter || has_default_setter)
                {
                    return Err(syn::Error::new(
                        ident.span(),
                        "A `#[property]` attribute which does not use any of the default `get` and `set` \
                        arguments -- e.g. #[property], #[property(get)] or #[property](get, set=\"path\") -- \
                        requires the field type to be `Property<T>`."
                    ));
                }

                // Field type is `Property<T>`
                if property_ty.is_some()
                    && (is_standalone_attribute || has_default_getter || has_default_setter)
                {
                    return Err(syn::Error::new(
                        ident.span(),
                        "The `#[property]` attribute requires explicit paths for `get` and `set` argument; \
                        the defaults #[property], #[property(get)] and #[property(set)] are not allowed."
                    ));
                }

                // if both of them are not set, i.e. `#[property]`. implicitly use both getter/setter
                let (get, set) = if is_standalone_attribute {
```

Reply:
I have got the same error message when running test. It occurs when I got `proc_macro::TokenStream` in a test function.

Reply:
It seems like I missed a `!` in my logic. For some reason, this doesn't just fail the test, but creates this entirely unrelated error message.